src/smif/data_layer/database_interface.py

```diff
@@ -72,19 +75,187 @@ def delete_sos_model(self, sos_model_name):
 
     # region Models
     def read_models(self):
-        raise NotImplementedError()
+        """Read all simulation models
+        """
+        # establish a cursor to read the database
+        cursor = self.database_connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+
+        # run sql call
+        cursor.execute('SELECT * FROM simulation_model')
+
+        # get returned data
+        simulation_models = cursor.fetchall()
+
+        # return data to user
+        return simulation_models
 
     def read_model(self, model_name):
-        raise NotImplementedError()
+        """Read a simulation model
+        """
+        # establish a cursor to read the database
+        cursor = self.database_connection.cursor(cursor_factory=psycopg2.extras.RealDictCursor)
+
+        # run sql call
+        cursor.execute('SELECT * FROM simulation_models WHERE name=%s', [model_name])
+
+        # get returned data
+        simulation_model = cursor.fetchall()
+
+        # check if a simulation model has been found and only proceed if so
+        if len(simulation_model) == 1:
+            # get port types for model - inputs, outputs, parameters
+            # run sql call
+            for port_type in self.port_types:
+                cursor.execute('SELECT s.* FROM simulation_model_port smp JOIN specifications s ON smp.specification_name=s.name WHERE smp.model_name=%s AND port_type=%s;', [model_name, port_type])
```

Comment:
I have changed methods around specifications so that they are referenced by their id once added, rather than always having to refer to model_name, port_type and spec_name.